staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go

```diff
@@ -126,7 +126,10 @@ func (w *watcher) createWatchChan(ctx context.Context, key string, rev int64, re
 		// The filter doesn't filter out any object.
 		wc.internalPred = storage.Everything
 	}
-	wc.ctx, wc.cancel = context.WithCancel(ctx)
+
+	// Wrap ctx with WithRequireLeader option to only succeed if the cluster has
+	// a leader to prevent hanging watches when the etcd majority is lost
+	wc.ctx, wc.cancel = context.WithCancel(clientv3.WithRequireLeader(ctx))
```

Comment:
I'm wondering what does it mean for leader reelections. Are we also going to break *all* watches?
Leader reelections may happen in perfectly healthy cluster (e.g. during rolling upgrade of masters), so I would like to ensure this is exactly what we want - what is the exact semantic of `WithRequireLeader`?

Another thing is that if this is infrequently changing resource type, it may trigger relists of all the watchers (even with watch-cache):
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L422
[that can potentially overload and kill control plane]

It's not a newly added thing with this PR, but it may make it happen much more frequently....

Reply:
> I'm wondering what does it mean for leader reelections. Are we also going to break all watches?

The etcd server waits until it cannot find a leader for 3 election timeouts to cancel existing streams. 3 is currently a hard coded constant. The election timeout defaults to 1000ms.

If the cluster is healthy, when the leader is stopped, the leadership transfer should be smooth. (leader transfers its leadership before stopping). If leader is hard killed, other servers will take an election timeout to realize leader lost and start campaign.

Reply:
OK - that sounds fine from my perspective.

Though it would be useful to add this explanation in the comment. @embano1 - can you please add comment like what @jingyih wrote?

Reply:
Updated the PR comment and the commit message with the explanation from @jingyih

Reply:
Please add this comment to this file - it's much easier to understand it than look for the PR that added it.

Reply:
@wojtek-t sorry, I misunderstood you then. With comment you mean in the code comment section (file) not the Github comment here, correct? Just wanted to clarify.

Reply:
Added comment to the file.

Reply:
Yes - sorry for not being clear.
And thanks for adding.